Updates to the NCO mode

danielabdi-noaa opened the pull request:
## DESCRIPTION OF CHANGES: 
<!-- One or more paragraphs describing the problem, solution, and required changes. -->
This PR is an update of NCO mode with the following changes
- Currently there is no way to install only the binaries in `HOMEdir` because dependencies like GSI do not have the option to install for example `include` in a separate directory from `bin`. A temporary solution thanks to @MatthewPyle-NOAA is to install in `build` directory and move the `exec` back to `$HOMEdir` so that the rest  `share`, `lib`, `lib64`, `include` do not pollute the `HOMEdir`.
   ```
   ./devbuild.sh -p=hera --install-dir=$PWD/build --move
   ```
   If you want to have some time to inspect before moving `exec` to `HOMEdir`, make this a two step-process by
   ```
   ./devbuild.sh -p=hera --install-dir=$PWD/build
   ./devbuild.sh -p=hera --install-dir=$PWD/build --continue --move
   ```
- Fix bug in calling setpdy without `COMROOT` set
- Fix bug in simultaneous run of post with forecast
- Remove use of `DATA_SHARED` directory and make all tasks use one temp directory each
- Make sure `KEEPDATA=FALSE` works properly. `$DATA` directory is now clean with the flag on.
- Fixes issue #442 by undoing the `for_ICS/LBCS` addition to NCO mode. The `COMIN` directory still has `$cyc`attached to it so should be fine for AQM?
- Add templated data source paths that use compath.py for wcoss2 (needs testing). The paths are set like this in `wcoss2.yaml`
   ```
   FV3GFS: compath.py ${envir}/gfs/${gfs_ver}/gfs.${PDY}
   ```
   I am hoping the template variable mechanism will work this way.
- First attempt at issue #445. Now about **40** test cases are covered compared to the previous **9**, and the tests should finish in about the same time. This is comparable to a comprhensive test IMO. Note that hera is running entirely in nco mode, so there are some repeat test cases. Let me know if you have suggestions on how to optimize this better.
- Potentially solves issue #349 . `get_extrn_ics/lbcs` tasks are assigned 1 core but there is no minimum memory requirement. Sometimes these tasks fail with `state DEAD (OUT_OF_MEMORY)` message, which I am hoping will be solved by adding a `<memory>4G</memory>` requirement to rocoto xml file. Comparing the second and third jet runs looks like this issue maybe solved.

### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 
<!-- Explicitly state what tests were run on these changes, or if any are still pending (for README or other text-only changes, just put "None required"). Make note of the compilers used, the platform/machine, and other relevant details as necessary. For more complicated changes, or those resulting in scientific changes, please be explicit! -->
<!-- Add an X to check off a box. -->

- [X] hera.intel (comprehensive test)
- [X] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [X] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [X] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:
<!-- Add any links to external PRs (e.g. regional_workflow and/or UFS PRs). For example:
- ufs-community/regional_workflow/pull/<pr_number>
- ufs-community/UFS_UTILS/pull/<pr_number>
- ufs-community/ufs-weather-model/pull/<pr_number> -->

## DOCUMENTATION:
<!-- If this PR is contributing new capabilities that need to be documented, please also include updates to the RST files (docs/UsersGuide/source) as supporting material. -->

## ISSUE: 
<!-- If this PR is resolving or referencing one or more issues, in this repository or elsewhere, list them here (Remember, issues must always be created before starting work on a PR branch!). For example, "Fixes issue mentioned in #123" or "Related to bug in https://github.com/ufs-community/other_repository/pull/63" -->

## CHECKLIST
<!-- Add an X to check off a box. -->
- [ ] My code follows the style guidelines in the Contributor's Guide
- [ ] I have performed a self-review of my own code using the Code Reviewer's Guide
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## LABELS (optional): 
<!-- If you do not have permissions to add labels to your own PR, request that labels be added here. 
Add an X to check off a box. Delete any unnecessary labels. -->
A Code Manager needs to add the following labels to this PR: 
- [ ] Work In Progress
- [ ] bug
- [ ] enhancement
- [ ] documentation
- [ ] release
- [ ] high priority
- [ ] run_ci
- [ ] run_we2e_fundamental_tests
- [ ] run_we2e_comprehensive_tests
- [ ] Needs Cheyenne test 
- [ ] Needs Jet test 
- [ ] Needs Hera test 
- [ ] Needs Orion test 
- [X] help wanted
     Since this PR also changes Jenkins (increases number of tests from 9 to 40), I need help in running Jenkins once to get info 
     on load balancing and test cases that fail (or not reliable) so that they can be removed from the list

## CONTRIBUTORS (optional): 
<!-- If others have contributed to this work aside from the PR author, list them here -->
@MatthewPyle-NOAA

danielabdi-noaa committed:
Add a move step to binaries for NCO purposes.

danielabdi-noaa committed:
Pass COMROOT to setpdy.

danielabdi-noaa committed:
Remove symlinks and use unique id for run_fcst instead.

danielabdi-noaa committed:
Remove DATA_SHARED directory.

danielabdi-noaa committed:
Correct path of post dependency on forecat.

danielabdi-noaa committed:
Remove forecast directory after post is done.

danielabdi-noaa committed:
Use raw workflow_id

danielabdi-noaa committed:
Undo for_ICS/LBCS stuff in NCO mode.

danielabdi-noaa committed:
Add templated data paths to wcoss2 that use compath.py.

danielabdi-noaa committed:
Remove cron relaunch from config.nco

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1103924331/20221029152015/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on hera: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_RRFS_v1beta
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
All experiments completed

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1103924331/20221029152014/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on jet: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_RRFS_v1beta
All experiments completed

danielabdi-noaa committed:
Remove RESTART symlinking because DATAFCST is accessible to all other tasks.

natalie-perlin commented:
These changes may need adaptation of the devclean.sh script.

danielabdi-noaa committed:
Update devclean.

danielabdi-noaa commented:
> These changes may need adaptation of the devclean.sh script.

Thanks, I've made sure `$HOMEdir/exec` is always cleaned up and made some minor modiffications in `devclean.sh`.

danielabdi-noaa committed:
Optimize fundamental test coverage.

danielabdi-noaa committed:
Follow symlinks when tarring log files in Jenkins.

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1103924331/20221030165012/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 10 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Failed on hera: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
2022-10-30 17:20:07 +0000 :: hfe06 :: Task make_ics, jobid=37242603, in state DEAD (FAILED), ran for 57.0 seconds, exit status=256, try=2 (of 2)
Experiment Succeeded on hera: pregen_grid_orog_sfc_climo
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on hera: community_ensemble_2mems_stoch
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Failed on hera: MET_ensemble_verification
2022-10-30 17:44:05 +0000 :: hfe01 :: Task run_gridstatvx_mem2, jobid=37243530, in state DEAD (FAILED), ran for 20.0 seconds, exit status=256, try=1 (of 1)
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
All experiments completed

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1103924331/20221030165013/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 10 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE
Experiment Failed on jet: specify_RESTART_INTERVAL
2022-10-30 17:32:09 +0000 :: fe2 :: Task get_extrn_ics, jobid=15049845, in state DEAD (OUT_OF_MEMORY), ran for 200.0 seconds, exit status=253, try=1 (of 1)
Experiment Failed on jet: specify_RESTART_INTERVAL
2022-10-30 17:32:09 +0000 :: fe2 :: Task get_extrn_lbcs, jobid=15049846, in state DEAD (OUT_OF_MEMORY), ran for 174.0 seconds, exit status=253, try=1 (of 1)
Experiment Failed on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
2022-10-30 17:28:14 +0000 :: fe4 :: Task get_extrn_lbcs, jobid=15049755, in state DEAD (OUT_OF_MEMORY), ran for 208.0 seconds, exit status=253, try=1 (of 1)
Experiment Failed on jet: custom_ESGgrid
2022-10-30 17:32:13 +0000 :: fe2 :: Task get_extrn_lbcs, jobid=15049853, in state DEAD (OUT_OF_MEMORY), ran for 110.0 seconds, exit status=253, try=1 (of 1)
Experiment Succeeded on jet: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on jet: specify_DT_ATMOS_LAYOUT_XY_BLOCKSIZE
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on jet: custom_GFDLgrid
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: specify_DOT_OR_USCORE
All experiments completed

danielabdi-noaa committed:
Add memory otpion for get_extrn_ics/lbcs.

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1103924331/20221030192016/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 10 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE
Experiment Succeeded on jet: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: custom_ESGgrid
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on jet: specify_DT_ATMOS_LAYOUT_XY_BLOCKSIZE
Experiment Succeeded on jet: specify_RESTART_INTERVAL
Experiment Succeeded on jet: custom_GFDLgrid
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on jet: specify_DOT_OR_USCORE
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
All experiments completed

danielabdi-noaa committed:
Undo cyc addition to COMIN.

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1103924331/20221030225013/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 10 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Failed on hera: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
2022-10-30 23:20:08 +0000 :: hfe05 :: Task make_ics, jobid=37251515, in state DEAD (FAILED), ran for 52.0 seconds, exit status=256, try=2 (of 2)
Experiment Succeeded on hera: community_ensemble_2mems_stoch
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on hera: pregen_grid_orog_sfc_climo
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on hera: MET_ensemble_verification
All experiments completed

danielabdi-noaa committed:
Reduce work load on orion.

danielabdi-noaa committed:
Add minimum memory requirement to all serial tasks.

danielabdi-noaa committed:
Simplify set_predef_grid_params.

danielabdi-noaa committed:
Simplify generate_FV3LAM_wflow.

danielabdi-noaa committed:
Simplify link_fix.

danielabdi-noaa committed:
Simplify to get rid of export_vars from setup.

danielabdi-noaa committed:
Use `eval echo` or `eval` depending on situation.

danielabdi-noaa committed:
Import only necessary variables from environment.

chan-hoo commented:
@danielabdi-noaa, I tried to run some WE2E tests for the nco mode, but I got the following error:   Reading in the test description for primary WE2E test:  "grid_CONUS_25km_GFDLgrid_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16"
  In category (subdirectory):  "wflow_features"

Traceback (most recent call last):
  File "/scratch2/NCEPDEV/stmp3/Chan-hoo.Jeon/srw_app_test/ufs-srweather-app/ush/calculate_cost.py", line 106, in <module>
    params = calculate_cost(args.cfg)
  File "/scratch2/NCEPDEV/stmp3/Chan-hoo.Jeon/srw_app_test/ufs-srweather-app/ush/calculate_cost.py", line 31, in calculate_cost
    DT_ATMOS,
**NameError: name 'DT_ATMOS' is not defined**

chan-hoo commented:
@danielabdi-noaa, I ran your branch with my sample script (not with WE2E). I found two issues: 
1) The `COMIN` directory doesn't have its `cyc` subdirectories although you mentioned they should be there in the above description. 
2) The reason why I created 'for_ICS/LBCS' in the 'get_extrn' tasks and deleted them in the 'make_ics/lbcs' tasks, the external data files, used as the input files for make_ics/lbcs, should not be in the `COMIN` directory. If these directories cause an issue for a specific case, you can remove them. However, the external data files (pgrb2, nemsio files, rap files) should be removed from the `COMIN`directory. For example, in the following COMIN directory, `gfs.t00z.pgrb2.0p25.f*` and `*.extrn_mdl_var_defns.sh` files should be removed:
(regional_workflow) [Chan-hoo.Jeon@hfe09 rrfs_test.20220920]$ ls
**gfs.t00z.pgrb2.0p25.f000**                      rrfs.t00z.natlev.f003.conus_25km.grib2
**gfs.t00z.pgrb2.0p25.f003**                      rrfs.t00z.natlev.f004.conus_25km.grib2
**gfs.t00z.pgrb2.0p25.f006**                      rrfs.t00z.natlev.f005.conus_25km.grib2
**rrfs.t00z.FV3GFS.ICS.extrn_mdl_var_defns.sh**   rrfs.t00z.natlev.f006.conus_25km.grib2
**rrfs.t00z.FV3GFS.LBCS.extrn_mdl_var_defns.sh**  rrfs.t00z.prslev.f000.conus_25km.grib2
rrfs.t00z.gfs_bndy.tile7.f000.nc              rrfs.t00z.prslev.f001.conus_25km.grib2
rrfs.t00z.gfs_bndy.tile7.f003.nc              rrfs.t00z.prslev.f002.conus_25km.grib2
rrfs.t00z.gfs_bndy.tile7.f006.nc              rrfs.t00z.prslev.f003.conus_25km.grib2
rrfs.t00z.gfs_ctrl.nc                         rrfs.t00z.prslev.f004.conus_25km.grib2
rrfs.t00z.gfs_data.tile7.halo0.nc             rrfs.t00z.prslev.f005.conus_25km.grib2
rrfs.t00z.natlev.f000.conus_25km.grib2        rrfs.t00z.prslev.f006.conus_25km.grib2
rrfs.t00z.natlev.f001.conus_25km.grib2        rrfs.t00z.sfc_data.tile7.halo0.nc
rrfs.t00z.natlev.f002.conus_25km.grib2

danielabdi-noaa committed:
Add cyc back to COMIN.

danielabdi-noaa committed:
Fix calculate_cost.

danielabdi-noaa commented:
> @danielabdi-noaa, I tried to run some WE2E tests for the nco mode, but I got the following error: Reading in the test description for primary WE2E test: "grid_CONUS_25km_GFDLgrid_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16" In category (subdirectory): "wflow_features"
> 
> Traceback (most recent call last): File "/scratch2/NCEPDEV/stmp3/Chan-hoo.Jeon/srw_app_test/ufs-srweather-app/ush/calculate_cost.py", line 106, in params = calculate_cost(args.cfg) File "/scratch2/NCEPDEV/stmp3/Chan-hoo.Jeon/srw_app_test/ufs-srweather-app/ush/calculate_cost.py", line 31, in calculate_cost DT_ATMOS, **NameError: name 'DT_ATMOS' is not defined**

@chan-hoo This is a bug introduced yesterday which I have fixed now.

danielabdi-noaa commented:
> @danielabdi-noaa, I ran your branch with my sample script (not with WE2E). I found two issues:
> 
> 1. The `COMIN` directory doesn't have its `cyc` subdirectories although you mentioned they should be there in the above description.

I've added that back in now. The MET verification tests were also failing before, but that turned out not to be related be your PR but to the fact that the `vx` steps are serial and do not specify memory. They can fail when that single core does not have enough memory -- has been stably running MET verification and other arbitrarily failing tests since I added minimum memory requirements.

> 2. The reason why I created 'for_ICS/LBCS' in the 'get_extrn' tasks and deleted them in the 'make_ics/lbcs' tasks, the external data files, used as the input files for make_ics/lbcs, should not be in the `COMIN` directory. If these directories cause an issue for a specific case, you can remove them. However, the external data files (pgrb2, nemsio files, rap files) should be removed from the `COMIN`directory. For example, in the following COMIN directory, `gfs.t00z.pgrb2.0p25.f*` and `*.extrn_mdl_var_defns.sh` files should be removed:

As I mentioned in the other PR ,there is a race condition happening when you run `community_ensemble_008mems` since all members use the same directory `for_ICS` to store their data. Maybe it will work if each member creates a unique folder that it deletes later, but deleting them causes any chance of re-running `make_ics`when it fails. From what I gathered from @MatthewPyle-NOAA , the COMIN directory can have anything it needs to run the forecast there, so those files do not need to be deleted since they allow re-runs.

chan-hoo commented:
@danielabdi-noaa, "From what I gathered from @MatthewPyle-NOAA , the COMIN directory can have anything it needs to run the forecast there, so those files do not need to be deleted since they allow re-runs." -> The external data files are not used to run the forecast. They are used only for make_ics/lbcs. As you know, the size of this file is quite big. When we archive the `COMIN` directory, we should remove these files because they are already on HPSS or other data stores.

danielabdi-noaa commented:
@chan-hoo I thought only the `grib2`files that RRFS produces are picked up from COMIN for archiving purposes, not the whole directory. In fact, we had also the forecast output (dyn/phy netcdf files in COMIN too) which are not there anymore. Re-running capability with minimal steps need to be executed is an NCO requirement (look page 13 item v), so if `make_ics` fails user would only need to re-run that step again if the gfs input files are there.

chan-hoo commented:
@danielabdi-noaa, you don't seem to agree with me :) @MatthewPyle-NOAA, can you confirm this? I don't understand why the external data files for make_ics/lbcs should be in the `COMIN` directory. If RRFS doesn't care about this, I'll remove them from Online-CMAQ only for AQM. However, I don't think this is what we want because Online-CMAQ will be merged into the UFS SRW App in the near future. Regarding dyn and phy files, I had to move them to `COMIN` for AQM separately.

JeffBeck-NOAA commented:
@MatthewPyle-NOAA

chan-hoo commented:
Thank you, @JeffBeck-NOAA! Updated.

chan-hoo commented:
@danielabdi-noaa , @MatthewPyle-NOAA, If @danielabdi-noaa's comment is true: "I thought only the grib2files that RRFS produces are picked up from COMIN for archiving purposes, not the whole directory. In fact, we had also the forecast output (dyn/phy netcdf files in COMIN too) which are not there anymore." => Can we put the 'dyn' and 'phy' files back to the `COMIN` directory? The AQM group wants them in the `COMIN` directory too.

danielabdi-noaa commented:
@chan-hoo Honestly, I do not agree even with the `${cyc}` addition to `COMIN` either because the standard is pretty clear that it should be part of the file name not a directory by itself. It is redundant because every file in COMIN has `$cyc` attached to it and is only added for AQM -- which for some reason needs special treatment.

> COMIN: com directory for current model's input data, typically  $COMROOT/$NET/$model_ver/$RUN.$PDY

The standard also tells you not to have `mem1` etc as a sub-directory but part of the file name.
 
> ensemble: atmospheric, hydro models: model.tHHz.ens_mem.var_info.f###.domain.format

So if AQM happens to have that, should RRFS then cater for that and add one more deviation from the NCO standard?

chan-hoo commented:
@danielabdi-noaa, please check pp.6 (13) "If the directory structure includes the cycle, it should be a subdirectory: <model>.YYYYMMDD/HH/". This cycle directory is essential for multiple cycle cases. In a multiple-cycle run, a next cycle should read some data in the previous cycle.

MichaelLueken commented:
@danielabdi-noaa The run_we2e_fundamental_tests label has been added to this PR to test the behavior of the modified, machine-based fundamental testing.

danielabdi-noaa commented:
@MichaelLueken Thanks that will give me some feedback on the load balancing! Orion ensemble test cases with 8 members especially takes long so would be interesting.

danielabdi-noaa commented:
@chan-hoo I read it as, if you choose to have cyc in the directory, not that you must do it. Particularly it doesn't want you to do `RUN.yyyymmddhh` but `RUN.yyyymmdd/hh`. If I recall correctly, @MatthewPyle-NOAA agreed to adding `$cyc` for AQM purposes, but it was not wrong the way we had it originally without it. There is no reason reason why AQM should not work without `$cyc` sub-directory as long as the file names are uniquely named... In any case, you have your `$cyc` sub-directory now :)

danielabdi-noaa committed:
Fix gaea test list.

chan-hoo commented:
@danielabdi-noaa, the UFS SRW App will be used by RRFS and AQM. RRFS doesn't care if it has the cycle directory (even dyn/phy files), but AQM requires it. In this situation, it will be more reasonable to put the cycle directory (or dyn/phy files) by default. I agree that the cycle directory is not necesssary for the current status of the UFS SRW App :) :)

MichaelLueken commented:
@danielabdi-noaa I believe the Orion tests have failed due to maintenance on the machine today.  I'll resubmit the tests once the machine is back.

MatthewPyle-NOAA commented:
Trying to catch up here.  There aren't hard and fast rules on whether a `cyc `subdirectory exists under `${RUN}.${PDY}`, but we had decided that we were good with it for RRFS since AQM wanted it.   And I agree with Chan-Hoo's statement here "The external data files are not used to run the forecast. They are used only for make_ics/lbcs"  The "external data" is expected to be picked up from elsewhere on disk - it doesn't need permanence within the RRFS COM space (may temporarily exist on disk under /tmp).

danielabdi-noaa commented:
@MichaelLueken Yes, I have run the same test manually on Orion yesterday just fine. Gaea had a typo in the machine list, which I have fixed now, so both Gaea and Orion need re-runs.

danielabdi-noaa commented:
> @danielabdi-noaa, the UFS SRW App will be used by RRFS and AQM. RRFS doesn't care if it has the cycle directory (even dyn/phy files), but AQM requires it. In this situation, it will be more reasonable to put the cycle directory (or dyn/phy files) by default. I agree that the cycle directory is not necesssary for the current status of the UFS SRW App :) :)

I doubt that it will ever be the case that you *must* use sub-directories for AQM as long as you name all files uniquely i.e. have `t${cyc}z`. It is a choice to have a *cyc* sub-directory for convenience  but not something you must do.

@MatthewPyle-NOAA They can go back to DATA but then again, get_extrn_ics/lbcs DATA directory need to be *shared* just like run forecast is without PID. If we go this route all DATA directories would need to be shared because a task would need something from the previous task anyway. If COMROOT is not the place to share anything, then all DATA directories should not have PID. Also @chan-hoo mentioned that he needs the forecast outputs `dyn/phy*.nc` to be in COMROOT for AQM, which is conflicting with the idea that we now want to move the gfs inputs back to DATA.

MatthewPyle-NOAA commented:
@danielabdi-noaa If there is a need for `dyn/phy*.nc` files to have permanence (used by a subsequent cycle of the same model, or used as an input by a different modeling system) they may end up in COM space.  But as a rule, what goes to COM needs a reason to be there (typically final output products for distribution, but also model IC/LBC/namelist files, and anything else needed to spin up production for a system after a production machine switch).  

The external data (say GFS information) is available from elsewhere on disk, and gets pulled in to the make_ic and make_bc jobs.  The make_ic and make_bc outputs go to COM for use by the forecast, but the GFS information is no longer needed.  I'm not understanding why this external data needs to exist in distinct shared directories under /tmp - can you clarify?

danielabdi-noaa commented:
> @danielabdi-noaa If there is a need for `dyn/phy*.nc` files to have permanence (used by a subsequent cycle of the same model, or used as an input by a different modeling system) they may end up in COM space. But as a rule, what goes to COM needs a reason to be there (typically final output products for distribution, but also model IC/LBC/namelist files, and anything else needed to spin up production for a system after a production machine switch).

That is also what I understood. The `gfs.` files are model ICS/LBCS so they can be in COMIN. We currently don't have namelist files (`input.nml`) in COMIN but I also thought it would abe Ok to have it there and anything you need for re-runs. For example `rrfs.t00z.FV3GFS.ICS.extrn_mdl_var_defns.sh` contains the list of GFS files and can not be deleted until `make_ics` successfully finishes.

> 
> The external data (say GFS information) is available from elsewhere on disk, and gets pulled in to the make_ic and make_bc jobs. The make_ic and make_bc outputs go to COM for use by the forecast, but the GFS information is no longer needed. I'm not understanding why this external data needs to exist in distinct shared directories under /tmp - can you clarify?

The GFS files are left there because it is Ok to have them since they are model input (COMIN) files. The `rrfs.*grib2` files are model outputs so they are in COMOUT, which by NCO standards overlaps with COMIN. The intermediate outputs of `make_ics` and `make_lbcs` is in COMIN because it is needed by forecast. So it seems that everything that is currently there in COMIN/OUT directory is Ok. However, if we want only grib2 files and something needed by the next cycle, we can clear them up after they are no longer needed like this:

- `gfs.*.grib2` input files, if they don't belong in COMIN, can be deleted after `make_ics/lbcs` successfully finish. If it fails, they don't get deleted so that we have re-run capability
- `rrfs.t${cyc}z.*.nc`  files (outputs of `make_ics/lbcs`) get deleted after `run_fcst` successfully finsihes.
- The forecast outputs if needed by the next cycle can remain in COMIN, so I can add those back in.

danielabdi-noaa committed:
Remove a test case from cheyenne that is taking extremely long.

danielabdi-noaa commented:
@MichaelLueken One test case on Cheyenne is taking extremely long
```
grid_SUBCONUS_Ind_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
```
It finishes quickly on Hera so I am not sure what is going on. In any case since we don't want cheyenne to take 2x longer, I've removed the particular test case. Could you please stop the test now so I can have the Jenkins log files? Maybe something has failed...

MichaelLueken commented:
@danielabdi-noaa I have stopped the Jenkins pipeline and resubmitted the Gaea tests.

MatthewPyle-NOAA commented:
> The GFS files are left there because it is Ok to have them since they are model input (COMIN) files. The rrfs.*grib2 files are model outputs so they are in COMOUT, which by NCO standards overlaps with COMIN. The intermediate outputs of make_ics and make_lbcs is in COMIN because it is needed by forecast. So it seems that everything that is currently there in COMIN/OUT directory is Ok. However, if we want only grib2 files and something needed by the next cycle, we can clear them up after they are no longer needed like this:

> gfs.*.grib2 input files, if they don't belong in COMIN, can be deleted after make_ics/lbcs successfully finish. If it fails, they don't get deleted so that we have re-run capability

@danielabdi-noaa An alternate idea on how to handle the external data.  Create a new subdirectory under the OPSROOT space to hold it.  Can name it whatever you want, and it can linger there.  This keeps it from being in the /tmp and /com kinds of spaces where we don't want it with operations.  In operations, these will be grabbed from COMINgfs.

> rrfs.t${cyc}z.*.nc files (outputs of make_ics/lbcs) get deleted after run_fcst successfully finsihes.

Definitely not on deleting the `rrfs.t${cyc}z.*.nc` files after the forecast successfully finishes.  Individual files aren't scrubbed from COMOUT in this way.  If something goes to COM, it is expected to stay there until that day of data ages off the system (different models retain a different number of days worth of data).  

> The forecast outputs if needed by the next cycle can remain in COMIN, so I can add those back in.

danielabdi-noaa commented:
@MichaelLueken Gaea C3 nodes have 2G/core limit according to the docs, but I was requesting for 4G so needs to reduce it to 1G just to be safe

> C3 is a Cray XC40 with 47872 Intel Haswell cores and 93.5TB of memory (2GB/core). It is available from all Gaea login nodes.

danielabdi-noaa committed:
Reduce memory requirement for serial job to 1G.

danielabdi-noaa commented:
@MatthewPyle-NOAA The `get_extrn_ics/lbcs` task uses a generic source `disk, aws, hpss` etc, so wherever it gets the data from it copies it to a local directory before being used for `make_ics/lbcs`. I thought `COMIN` is the perfect place for that, given that it is the place to store model input data. So if I understand your suggestion, on wcoss2, `COMINgfs` data should not be copied, and I should store it somewhere shared under for example `OPSROOT/shared`. A symlink to COMINgfs file will work too, but if the source is `aws`, I would need to copy it.

MichaelLueken commented:
@danielabdi-noaa The Gaea Jenkins CI failed again this morning.  Running Orion pipeline now since the machine is out of maintenance.

danielabdi-noaa commented:
@MichaelLueken Thanks. The error still says memory specification can not be satsified
```
11/03/22 10:04:22 EDT :: FV3LAM_wflow.xml :: Submission of get_extrn_ics failed!  sbatch: error: Memory specification can not be satisfied
sbatch: error: Batch job submission failed: Requested node configuration is not available
```
I have no idea why that is given that gaeadocs says that 2Gb/core is Ok.

christopherwharrop-noaa commented:
@danielabdi-noaa - Why are you designating a memory request for a non-serial job? Is this intended to be a job that shares nodes with other jobs?

danielabdi-noaa commented:
> 

@christopherwharrop-noaa I am requesting minimum memory only for **serial** jobs. Slurm is sometimes assigning no memory for serial jobs and they fail arbitrarily because of that. If you look above to some of the github actions runs, before addinn the requirement I had four failures on Jet with `in state DEAD (OUT_OF_MEMORY)` and afterwards they are all gone.

christopherwharrop-noaa commented:
@danielabdi-noaa - Ok.  When I looked, it appeared that 'get_extrn_ics` was not a serial job.  So that was why I was asking.

MichaelLueken commented:
@danielabdi-noaa In the past, I've encountered failures due to Slurm not allocating memory to nodes with the GSI.  Rather than specifying the amount of memory to use, I added the following:

`#SBATCH --mem=0`

Would using this method work better here?

MichaelLueken commented:
@danielabdi-noaa The Orion Jenkins pipeline has ultimately failed, with five tests passing and three tests failing.  The three failures are:

```
grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta
GST_release_public_v1
MET_verification
```

`grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta` failed with:
```
 FATAL ERROR:
 OPENING: /work/noaa/epic-ps/role-epic-ps/jenkins/workspace/fs-srweather-app_pip
 eline_PR-443/expt_dirs/grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta/
 fix_lam/C403.slope_type.tile7.nc: No such file or directory
 STOP.
```
`GST_release_public_v1` test failed because the walltime limit was reached.

`MET_verification` failed with:
`/usr/bin/env: python3: No such file or directory`

The system is sluggish, which could have led to the walltime limit failure for `GST_release_public_v1`, but I'm unsure about the other two failures.  The C403.slope_type.tile7.nc file is present and the C403.slope_type.tile7.halo4.nc file in sfc_climo is present and not an empty file.  Why was the task unable to open the file when it is clearly present?  All tasks successfully ran for the `MET_verification` test, except for the verification tasks themselves.  Why would the verification jobs fail due to python3 while the rest of the tasks ran without issue?

MichaelLueken reviewed (commented):
@danielabdi-noaa While I'm fine with the changes that have been made in this PR, there are still issues with the reworked fundamental tests.  Once the fundamental tests run successfully on Gaea and Orion, I will approve these changes.  Also, I would like to see @chan-hoo's approval before these changes are merged.

danielabdi-noaa commented:
@MichaelLueken Thanks for looking into the issues -- just got time to look into them now. One thing that is different from my local testing on Orion is the new EPIC installed modulefiles -- I used the old ones that loaded python3. I don't have access to test on gaea but it would be good to know what `--mem=x` it likes by submitting an `srun -M c3 --mem=x` command  -- the last resort will be to disable it for gaea or use `--mem=0` as you suggested. 

About the NCO mode we have discussed what needs to be done today, and I will be making a commit to address that.

Thanks

danielabdi-noaa committed:
Create COMINaws ics/lbcs staging directory.

danielabdi-noaa committed:
Also symlink gfs ics/lbcs if on disk and in community mode.

danielabdi-noaa committed:
Remove failing tests on orion.

danielabdi-noaa committed:
Exclude gaea from --mem specification.

danielabdi-noaa committed:
Add a run_vx modulefile for orion.

danielabdi-noaa committed:
Add separate fundamental list for cheyenne GNU runs.

MatthewPyle-NOAA commented:
@danielabdi-noaa Noticed an issue on WCOSS2 in machine/wcoss2.yaml.  

`  FV3GFS: compath.py ${envir}/gfs/${fv3gfs_ver}/gfs.${PDY}
`

fv3gfs_ver isn't defined in the WCOSS2 versions file.  Could you change fv3gfs_ver to gfs_ver here?  

A longer term goal might be to change all FV3GFS references to GFS.  This distinction might have made sense shortly after the GFS began running with the FV3 model, but shouldn't be necessary now.

danielabdi-noaa committed:
Bug fix for wcoss2 GFS version

MichaelLueken commented:
@danielabdi-noaa The Jenkins CI tests on Gaea are now passing.  I wish I knew why the machine wasn't accepting values for `--mem=x`.  I submitted several tests yesterday afternoon and this morning.  The only value that worked was using 0 (including 0G).  This removed the `--mem` option from the launcher.

I will launch the Orion tests, then the Cheyenne tests.

danielabdi-noaa commented:
@MichaelLueken Yes, I think `0G` allocates all available memory on the node to the serial task, but most likely shared rather than exclusive. It has these limits though so may not be the entire thing.
```
DefMemPerCPU=2048 MaxMemPerCPU=4096
```

christinaholtNOAA commented on scripts/exregional_run_fcst.sh:
```diff
@@ -515,21 +487,6 @@ POST_STEP
-fi
-#
-#-----------------------------------------------------------------------
-#
```
This is the part that means we are leaving these files in place instead of putting them in COM, right?

danielabdi-noaa replied:
Yes. Forecast outputs remain in `DATAROOT/run_fcst.unique_id` but are accessible since this directory is shared. The RESTART file is also in there and is also accessible if it is needed by later cycles.

christinaholtNOAA commented on tests/WE2E/machine_suites/fundamental.cheyenne:
```diff
@@ -0,0 +1,9 @@
+grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
+pregen_grid_orog_sfc_climo
+custom_GFDLgrid__GFDLgrid_USE_NUM_CELLS_IN_FILENAMES_eq_FALSE
+custom_GFDLgrid__GFDLgrid_USE_NUM_CELLS_IN_FILENAMES_eq_TRUE
```
These added machine_suite files seem outside the scope of NCO compliance. What am I missing?

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -108,6 +108,13 @@ def generate_FV3LAM_wflow(USHdir, logfile: str = "log.generate_FV3LAM_wflow") ->
+        settings = {}
+        for k, v in var_defs_dict.items():
+            settings[lowercase(k)] = v
+
```
YES!

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -893,44 +904,27 @@ def get_location(xcs, fmt):
-        MAILCC
-    except NameError:
+    if MAILCC is None:
         MAILCC = None
```
Are the last two needed?

danielabdi-noaa replied:
I've removed them.

christinaholtNOAA commented on ush/job_preamble.sh:
```diff
@@ -104,9 +114,28 @@ export -f POST_STEP
 if [ "${RUN_ENVIR}" = "nco" ]; then
     export COMIN="${COMIN_BASEDIR}/${RUN}.${PDY}/${cyc}"
     export COMOUT="${COMOUT_BASEDIR}/${RUN}.${PDY}/${cyc}"
+    export COMINaws="${AWSROOT}/${RUN}.${PDY}/${cyc}"
```
I'm curious what AWSROOT is? Is it already documented somewhere? Is this AWS different from Amazon Web Services?

MatthewPyle-NOAA replied:
> I'm curious what AWSROOT is? Is it already documented somewhere? Is this AWS different from Amazon Web Services?

It is being used as a holding spot for input files being brought into the system.  Has to do with some of the discussion/chatter recently about the GFS files not belonging in RRFS com space. 
Maybe COMINexternal/EXTERNALROOT would be more general,  and less likely to cause confusion?

danielabdi-noaa replied:
I will rename them to COMINext, EXTROOT and OPSROOT/ext if that is Ok? When we use wcoss2 sources COMINgfs, COMINhrrr etc, EXTROOT will just symlink the files and is used as a place to communicate the ICS/LBCS data to the `make_ics/lbcs`.

MatthewPyle-NOAA replied:
That sounds good to me, @danielabdi-noaa

christinaholtNOAA replied:
I think that is a nicer alternative given that we would certainly have external sources that are not aws. Thanks @danielabdi-noaa and @MatthewPyle-NOAA

christinaholtNOAA reviewed (approved):
@danielabdi-noaa I think the code looks good. 

I left a few minor comments below. And a couple of clarification questions. Nothing that should hold up progress, though.

MatthewPyle-NOAA commented:
More changes are needed for WCOSS2  - use of compath.py throws an error since prod_util isn't being loaded.  Can you:

Add `export prod_util_ver=2.0.14`  to the versions/run.ver.wcoss2 file?

And use it in in the modulefiles/tasks/wcoss2/*.lua definitions?  

And another item that seems to have been an on again, off again issue is that GRIB output isn't going to com as it is generated.

danielabdi-noaa committed:
Load prod_util on wcoss2.

danielabdi-noaa committed:
Rename AWSROOT to EXTROOT.

danielabdi-noaa committed:
Remove leftover COMINaws.

danielabdi-noaa commented:
@MichaelLueken Orion is currently under maintenance and I think the tests are stuck so best to stop them. In any case, those 5 test cases are ones who successfully run on orion before, so they should work once orion maintenance is over.

MichaelLueken commented:
@danielabdi-noaa I've been testing nco mode on Hera manually.  Once the tests there pass, I will approve (they had been failing due to COMINaws remnants in the preamble).

MichaelLueken reviewed (approved):
@danielabdi-noaa The manual tests on Hera successfully passed. I am now approving this PR.

danielabdi-noaa commented:
@MichaelLueken I have run the comprhensive test case (83 of them) on Hera in NCO mode. All have exceeded except 4 which are known failures. @MatthewPyle-NOAA The grib2 files are all in COMIN but I do recall having the issue you mentioned when post processing failed. Here is the OPSROOT directory on Hera for the new run
```
/scratch1/BMC/zrtrr/Daniel.Abdi/rrfs/OPSROOT-nco-2
```
Let me know if that still happens. I am going to merge this now.

mkavulich commented on tests/WE2E/run_WE2E_tests.sh:
```diff
@@ -430,10 +430,24 @@ elif [ -n "${tests_file}" ] || [ -n "${test_type}" ] ; then
   if [ -n "${test_type}" ] ; then
-    # Check for a pre-defined set. It could be machine dependent or not.
-    user_spec_tests_fp=${scrfunc_dir}/machine_suites/${test_type}.${machine}
+    # Check for a pre-defined set. It could be machine dependent or has the mode
+    # (community or nco), or default
+    user_spec_tests_fp=${scrfunc_dir}/machine_suites/${test_type}.${machine}.nco
     if [ ! -f ${user_spec_tests_fp} ]; then
-        user_spec_tests_fp=${scrfunc_dir}/machine_suites/${test_type}
+        user_spec_tests_fp=${scrfunc_dir}/machine_suites/${test_type}.${machine}.com
+        if [ ! -f ${user_spec_tests_fp} ]; then
+            user_spec_tests_fp=${scrfunc_dir}/machine_suites/${test_type}.${machine}.${compiler}
+            if [ ! -f ${user_spec_tests_fp} ]; then
+                user_spec_tests_fp=${scrfunc_dir}/machine_suites/${test_type}.${machine}
+                if [ ! -f ${user_spec_tests_fp} ]; then
+                    user_spec_tests_fp=${scrfunc_dir}/machine_suites/${test_type}
+                fi
+            fi
+        else
+            run_envir=${run_envir:-"community"}
+        fi
+    else
+        run_envir=${run_envir:-"nco"}
```
@danielabdi-noaa I hadn't checked on this PR before it was merged because I thought it would only change NCO mode running, but this seems to have introduced the additional change that the fundamental test suite on Hera will *always* be run in NCO mode. Was this an intentional change? If so, what was the reasoning?

danielabdi-noaa replied:
@mkavulich Yes, it is intentional. The goal for the fundamental test suites has shifted with this PR, now it is more about expanded test coverage and variety. I chose Hera for NCO mode so that it would be tested with the Github Actions CI as well, which runs only on Hera/Jet. The fundamental test suite is now running different test cases on each platform, each compiler gnu/intel. Last time I check it was close to being "comprehensive" with 45 test cases or so.

The logic in this code is kind of convoluted and if you find a better way to rename the files while keeping the flexibility, that would be great!

mkavulich replied:
Thanks for getting back to me so quickly. I may try to bring this up for discussion at the next code management meeting, because this feels like it may be confusing to developers testing their changes on a specific machine. Although perhaps a better topic for discussion would be the broader "which WE2E tests should developers be running to test their code prior to opening a PR", since this is not yet covered in the contributors guide, and clearly I didn't even understand what tests were being run when I ran the script before now!
